nup/update.go

```diff
@@ -116,6 +116,22 @@ func (u Update[T]) Apply(value T) T {
 	}
 }
 
+// ApplyPtr returns the result of applying the update to the given pointer
+// value. The result is the given value if the update is a no-op, nil if it's a
+// removal, or a copy of the update's contained value if it's a set operation.
+func (u Update[T]) ApplyPtr(value *T) *T {
+	switch u.op {
+	case OpNoop:
+		return value
```

Comment:
That's a good question. My reasoning is that `value` comes from the caller, so they already have full "ownership" of it and can mutate/copy it if they want to. If they need to mutate the result of `ApplyPtr` without mutating the argument they passed in, they have the option to make a deep copy on their side. On the other hand, we have to make a copy when returning the update's value if we want to prevent them from monkeying with private data. (Ideally, I'd like to return a pointer to an immutable object so we get both efficiency and proper encapsulation, but Go doesn't support that.)

For what it's worth, [here's the usage](https://github.com/nicheinc/fact/blob/139dac496c6a09c34f702ea774d458623cd8a0a2/validation/validator.go#L604-L606) of this method in https://github.com/nicheinc/fact/pull/301. In this case, the pointee is never mutated, and would I expect that to be the typical pattern.

Reply:
I'm good with this, and the behavior matches the documentation

> The result is the given value